include/circt/Dialect/Moore/MooreOps.td

```diff
@@ -199,9 +306,607 @@ def NonBlockingAssignOp : AssignOpBase<"nonblocking_assign"> {
+  }]>
+]> {
+  let description = [{
+    Compares the left- and right-hand side operand and returns a single bit 0,
+    1, or X result. If any bit in the two operands is Z or X, returns X.
+    Otherwise, if all bits are 0 or 1, `lt`, `le`, `gt`, and `ge` return whether
+    the left-hand side is less than, less than or equal to, greater than, or
+    greater than or equal to the right-hand side, respectively. `lt` corresponds
+    to the `<` operator, `le` to `<=`, `gt` to `>`, and `ge` to `>=`.
+
+    See IEEE 1800-2017 § 11.4.4 "Relational operators".
+  }];
+  let arguments = (ins SimpleBitVectorType:$lhs, SimpleBitVectorType:$rhs);
+  let results = (outs SingleBitType:$result);
+  let assemblyFormat = [{
+    $lhs `,` $rhs attr-dict `:` type($lhs) `->` type($result)
+  }];
+}
+
+def LtOp : RelationalOpBase<"lt"> { let summary = "Less than"; }
+def LeOp : RelationalOpBase<"le"> { let summary = "Less than or equal"; }
+def GtOp : RelationalOpBase<"gt"> { let summary = "Greater than"; }
+def GeOp : RelationalOpBase<"ge"> { let summary = "Greater than or equal"; }
+
+def ConcatOp : MooreOp<"concat", [
+    Pure, DeclareOpInterfaceMethods<InferTypeOpInterface>
+]> {
+  let summary = "A concatenation of expressions";
+  let description = [{
+    This operation represents the SystemVerilog concatenation expression
+    `{x, y, z}`. See IEEE 1800-2017 §11.4.12 "Concatenation operators".
+
+    All operands must be simple bit vector types.
+
+    The concatenation result is a simple bit vector type. The result is unsigned
+    regardless of the sign of the operands (see concatenation-specific rules in
+    IEEE 1800-2017 §11.8.1 "Rules for expression types"). The size of the result
+    is the sum of the sizes of all operands. If any of the operands is
+    four-valued, the result is four-valued; otherwise it is two-valued.
+  }];
+  let arguments = (ins Variadic<SimpleBitVectorType>:$values);
+  let results = (outs SimpleBitVectorType:$result);
+  let assemblyFormat = [{
+    $values attr-dict `:` functional-type($values, $result)
+  }];
+}
+
+def ReplicateOp : MooreOp<"replicate", [
+  Pure
+]> {
+  let summary = "Multiple concatenation of expressions";
+  let description = [{
+    This operation indicates a joining together of that many copies of the 
+    concatenation `{constant{w}}`. Which enclosed together within brace.
+    The 'constant' must a non-negative, non-x, and non-z constant expression.
+    The 'constant' may be a value of zero, but it only exists in parameterized
+    code, and it will be ignored(type is changed to the void).
+    Example:
+    ```
+      {4{w}}   // the same as {w, w, w, w}
+    ```
+    See IEEE 1800-2017 §11.4.12 "Concatenation operators".
+  }];
+  let arguments = (ins SimpleBitVectorType:$value);
+  let results = (outs SimpleBitVectorType:$result);
+  let assemblyFormat = [{
+    $value attr-dict `:` functional-type($value, $result)
+  }];
+}
+
+def ExtractOp : MooreOp<"extract"> {
```

Comment:
A new op to handle bits selections.

Reply:
Very cool! `ExtractOp` and `ReplicateOp` look very useful. Could you open a separate PR for them directly upstream in `llvm/circt`? Together with the corresponding parsing code and tests in `basic.sv` and `test/Dialect/Moore/basic.mlir`. :partying_face:

Reply:
No problem! I'll create a new PR upstream later.